src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,309 @@
+    network_client = get_network_client()
+    try:
+        network_client.network_interfaces.begin_create_or_update(
+            resource_group, nic.name, nic
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "associate NSG with NIC had conflicts",
+                "with concurrent request, ignoring %s",
+                err,
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "Unable to associate nsg %s with nic %s due to %s"
+                % (
+                    name,
+                    nic.name,
+                    err,
+                )
+            ],
+        )
+
+    return None
+
+
+def dissociate_nic(name: str, nic: NetworkInterface) -> Union[None, Error]:
+    if nic.network_security_group is None:
+        return None
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+    if nsg.id != nic.network_security_group.id:
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "network interface is not associated with this nsg.",
+                "nsg %s, nic: %s, nic.nsg: %s"
+                % (
+                    nsg.id,
+                    nic.name,
+                    nic.network_security_group.id,
+                ),
+            ],
+        )
+
+    logging.info("dissociating nic %s with nsg: %s %s", nic.name, resource_group, name)
+
+    nic.network_security_group = None
+    network_client = get_network_client()
+    try:
+        network_client.network_interfaces.begin_create_or_update(
+            resource_group, nic.name, nic
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "associate NSG with NIC had conflicts with ",
+                "concurrent request, ignoring %s",
+                err,
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "Unable to associate nsg %s with nic %s due to %s"
+                % (
+                    name,
+                    nic.name,
+                    err,
+                )
+            ],
+        )
+
+    return None
+
+
+class NSG(BaseModel):
+    name: str
+    region: Region
+
+    @validator("name", allow_reuse=True)
+    def check_name(cls, value: str) -> str:
+        if len(value) > 80:
+            raise ValueError("NSG name too long")
+        return value
+
+    def create(self) -> Union[None, Error]:
+        if self.get() is not None:
```

Comment:
I think we're using `get()` as an existence check and skipping the create call if it succeeds, but this warrants a comment.

Reply:
Important to note: is this existence check necessary to avoid an error? Is it an optimization? &c